docs/IR.md

```diff
@@ -1,127 +1,382 @@
+
+There are two distinct ways to pass information to operators – inputs and attributes. The latter are used for values that are literal constants in the graph, while the former represent graph inputs or values computed elsewhere in the graph. This distinction may be highly relevant to achieving good performance for some implementations, while completely irrelevant to others.
 
 ### Graphs
 
-Each computation dataflow graph is structured as a list of nodes that form a graph. Each node is a call to an operator. Nodes have zero or more inputs, one or more outputs, and zero or more attribute-value pairs. Additionally, a valid ONNX graph must fulfill the following requirements:
+A serialized graph is comprised of a set of metadata fields, a list of model parameters, and a list of computation nodes.
+
+Each computation dataflow graph is structured as a topologically sorted list of nodes that form a graph, which MUST be free of cycles. Each node represents a call to an operator. Each node has zero or more inputs and one or more outputs.
+
+Graphs have the following properties:
 
-- The graph MUST be free of cycles.
-- The graph MUST be in SSA (static single assignment) form, meaning outputs of all nodes are unique.
-- The nodes list MUST be in topologically sorted order, meaning if input of node `N2` is the output of node `N1`, `N2` must appear after `N1` in the nodes list.
+|Name|Type|Description|
+|---|---|---|
+name|string|The name of the model graph.
+node|Node[]|A list of nodes, forming a partially ordered computation graph based on input/output data dependencies.
+initializer|Tensor[]|A list of named tensor values, used to specify default values for some of the inputs of the graph. Each initializer value is associated with an input by name matching.
+doc_string|string|A human-readable documentation for this model. Markdown is allowed.
+input|ValueInfo[]|The input “parameters” of the graph, possibly initialized by a default value found in ‘initializer.’
+output|ValueInfo[]|The output parameters of the graph. Once all output parameters have been written to by a graph execution, the execution is complete.
+value_info|ValueInfo|__TODO: Write this description__
 
-#### Model Graph Metadata
+Each graph MUST define the names and types of its inputs and outputs, which are specified as ‘value info’ structures, having the following properties:
 
-The following describes the metadata properties of a model graph:
+Name|Type|Description
+|---|---|---|
+name|string|The name of the value/parameter.
+type|Type|The type of the value.
+doc_string|string|A human-readable documentation for this value. Markdown is allowed.
 
-|Name|Type|Format|Description|
-|----|----|------|-----------|
-|name|string|Valid C identifier|A name for the model.|
-|doc_string|string|Free form|A human-readable documentation string intended to summarize the purpose of the model. Markdown is allowed.|
+Each graph MUST specify a name.
```

Comment:
Why do we make graph's name required?

Reply:
Runtimes are not the only consumers of models; some tools will need something to identify the model. It's also important to identify models in logging, error messages, etc. In the future, if and when we get to supporting model composition, names will be essential for that ability, too.

Reply:
But we already have the model name, I am totally confused why we need both graph name and model name. Especially, each model can only contain one graph...

Reply:
I don't find a name property in ModelProto. The model name is stored in the graph message.

From the current onnx.in.proto, commenting on the domain name of a model:

  // Domain name of the model.
  // We use reverse domain names as name space indicators. For example:
  // `com.facebook.fair` or `com.microsoft.cognitiveservices`
  //
  // Together with `model_version` and GraphProto.name, this forms the unique identity of
  // the graph.
  optional string domain = 4;